### PR TITLE
Adds upgrade worker to model agent

### PR DESCRIPTION
**Pull request by tlm**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

*Please replace with a description about why this change is needed, along with a description of what changed?*

## QA steps

*Please replace with how we can verify that the change works?*

```sh
cd tests && ./main.sh caasadmission
```

2nd
```sh
export DOCKER_USERNAME=<your-dockerhub-username>
make install operator-image microk8s-operator-update push-operator-image
juju bootstrap microk8s
# Check that the model agent has come up in the controller namespace
juju add-model test
# Check that the model agent has come up in the test namespace

export JUJU_BUILD_NUMBER=1
make install operator-image microk8s-operator-update push-operator-image
juju upgrade-controller --agent-stream=develop --agent-version=2.8.1.1
# Wait for controller to restart and verify that the new controller that comes up has the new image and version. The model agent has the correct version and image.
```

## Documentation changes

*Please replace with any notes about how it affects current user workflow? CLI? API?*

## Bug reference

*Please add a link to any bugs that this change is related to.*

### CODE REVIEW

File: caas/kubernetes/provider/upgrade.go [hunk 1]
```diff
@@ -17,19 +18,118 @@ import (
...
+
+func upgradeAgent(agentTag string, vers version.Number) error {
+	tag, err := names.ParseTag(agentTag)
+	if err != nil {
+		return errors.Annotate(err, "parsing agent tag to upgrade")
+	}
+
+	name := tag.Id()
+	switch tag.Kind() {
+	case names.MachineTagKind:
```
Comment (wallyworld) [inline caas/kubernetes/provider/upgrade.go:57]: there should never be a machine tag for k8s

Comment (tlm) [reply caas/kubernetes/provider/upgrade.go:28]: This is just keeping the logic that was already there.

File: caas/kubernetes/provider/controller_upgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	c.Assert(ss.Spec.Template.Spec.Containers[0].Image, gc.Equals, newImagePath)
+
+	c.Assert(ss.Annotations[labelVersion], gc.Equals, version.MustParse("9.9.9").String())
+	c.Assert(ss.Spec.Template.Annotations[labelVersion], gc.Equals, version.MustParse("9.9.9").String())
+}
+
+func (s *ControllerUpgraderSuite) TestControllerDoesNotExist(c *gc.C) {
+	var (
+		appName = JujuControllerStackName
+	)
```
Comment (wallyworld) [inline caas/kubernetes/provider/controller_upgrade_test.go:91]: do we need this var?

File: caas/kubernetes/provider/modeloperator_upgrade.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+func (u *upgradeCAASModelOperatorBridge) Namespace() string {
+	return u.namespaceFn()
+}
+
+func (k *kubernetesClient) upgradeModelOperator(agentTag names.Tag, vers version.Number) error {
+	broker := &upgradeCAASModelOperatorBridge{
+		clientFn:    k.client,
+		namespaceFn: k.GetCurrentNamespace,
+	}
+	return modelOperatorUpgrade(modelOperatorName, vers, broker)
```
Comment (wallyworld) [inline caas/kubernetes/provider/modeloperator_upgrade.go:47]: seems overkill for a one line method?
just do

	return upgradeDeployment(operatorName, "", vers,
		broker.Client().AppsV1().Deployments(broker.Namespace()))

?

File: caas/kubernetes/provider/controller_upgrade.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+func controllerUpgrade(appName string, vers version.Number, broker UpgradeCAASControllerBroker) error {
+	return upgradeStatefulSet(appName, "", vers, broker.Client().AppsV1().StatefulSets(broker.Namespace()))
+}
+
+func (k *kubernetesClient) upgradeController(agentTag names.Tag, vers version.Number) error {
+	broker := &upgradeCAASControllerBridge{
+		clientFn:    k.client,
+		namespaceFn: k.GetCurrentNamespace,
+	}
+	return controllerUpgrade(bootstrap.ControllerModelName, vers, broker)
```
Comment (wallyworld) [inline caas/kubernetes/provider/controller_upgrade.go:45]: overkill for a one line method?
just do

return upgradeStatefulSet(appName, "", vers, broker.Client().AppsV1().StatefulSets(broker.Namespace()))

?

File: caas/kubernetes/provider/controller_upgrade_test.go (hunk 2 above)
Comment (tlm) [reply caas/kubernetes/provider/controller_upgrade_test.go:91]: I think so

File: caas/kubernetes/provider/modeloperator_upgrade.go (hunk 3 above)
Comment (tlm) [reply caas/kubernetes/provider/modeloperator_upgrade.go:47]: The reason is testing. Which I forgot to include.

File: cloudconfig/podcfg/image.go [hunk 5]
```diff
@@ -13,9 +13,9 @@ import (
 )
 
 const (
-	jujudOCINamespace = "jujusolutions"
-	jujudOCIName      = "jujud-operator"
-	jujudbOCIName     = "juju-db"
+	JujudOCINamespace = "jujusolutions"
+	JujudOCIName      = "jujud-operator"
+	JujudbOCIName     = "juju-db"
```
[the remark this answers is not in the document]
Comment (tlm) [reply cloudconfig/podcfg/image.go:18]: The tests wouldn't break if we changed these. I am just trying to keep it consistent. Still want them changed?

File: caas/kubernetes/provider/controller_upgrade.go (hunk 4 above)
Comment (tlm) [reply caas/kubernetes/provider/controller_upgrade.go:47]: it's for testing and it's not tied to the lifecycle of the kubernetes client object then.

**Comment by tlm** [comment]
$$merge$$

**Comment by manadart** [comment]
Merged without a description?

**Pull request merged**